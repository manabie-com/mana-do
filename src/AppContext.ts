import { createContext, useContext } from 'react';
import {AppState, initialState} from "./store/reducer";

export type GlobalContextType = {
  state: AppState
  dispatch: Function
}

export const useAppContext = () => useContext<GlobalContextType>(AppContext);

// @ts-ignore
const AppContext = createContext<GlobalContextType>(initialState);

export default AppContext;
