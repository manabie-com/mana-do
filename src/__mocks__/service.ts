import service from "service/api-frontend";

export default service;
