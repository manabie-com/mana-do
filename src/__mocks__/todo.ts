import { TodoStatus } from '@/models/todo'
import { EnhanceTodoStatus } from '@/pages/todo'

export const stateProvider = {
  todos: [
    {
      content: 'Meet some cute peoples at Manabie',
      created_date: '',
      status: TodoStatus.ACTIVE,
      id: '_1',
      user_id: 'aldenn'
    },
    {
      content: 'Work with Manabie',
      created_date: '',
      status: TodoStatus.ACTIVE,
      id: '_2',
      user_id: 'aldenn'
    },
    {
      content: 'Complete the test for todo app',
      created_date: '',
      status: TodoStatus.COMPLETED,
      id: '_3',
      user_id: 'aldenn'
    }
  ]
}

export const statusesFilterProvider: {
  description: string
  status: EnhanceTodoStatus
  numOfRecord: number
}[] = [
  {
    description: 'The number of active todo',
    status: TodoStatus.ACTIVE,
    numOfRecord: 2
  },
  {
    description: 'The number of completed todo',
    status: TodoStatus.COMPLETED,
    numOfRecord: 1
  },
  {
    description: 'All of todo item',
    status: 'ALL',
    numOfRecord: 3
  }
]

export const emptyLabelByStatusProvider: {
  description: string
  status: EnhanceTodoStatus
  label: string
}[] = [
  {
    description:
      "The empty label for ACTIVE status should be Nothing need to do, let's relax",
    status: TodoStatus.ACTIVE,
    label: "Nothing need to do, let's relax"
  },
  {
    description:
      'The empty label for COMPLETED status should be Nothing is completed',
    status: TodoStatus.COMPLETED,
    label: 'Nothing is completed'
  },
  {
    description:
      'The empty label for ALL status should be Type something into the input and Enter',
    status: 'ALL',
    label: 'Type something into the input and Enter'
  }
]
