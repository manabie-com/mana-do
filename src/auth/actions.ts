export const REQUEST = '@Auth/Request'
export const REQUEST_FAIL = '@Auth/Fail'
export const AUTHENTICATE_SUCCESS = '@Auth/AuthenticateSuccess'
export const LOGOUT_SUCCESS = '@Auth/LogoutSuccess'

interface RequestAction {
  type: typeof REQUEST,
}

interface RequestFailAction {
  type: typeof REQUEST_FAIL,
  error: string,
}

interface AuthenticateSuccessAction {
  type: typeof AUTHENTICATE_SUCCESS
  token: string
}

interface LogoutSuccessAction {
  type: typeof LOGOUT_SUCCESS
}

export function request(): RequestAction {
  return { type: REQUEST }
}

export function requestFail(error: string): RequestFailAction {
  return { type: REQUEST_FAIL,error }
}

export function authenticateSuccess(token: string): AuthenticateSuccessAction {
  return { type: AUTHENTICATE_SUCCESS, token }
}

export function logoutSuccess(): LogoutSuccessAction {
  return { type: LOGOUT_SUCCESS }
}

export type Action = RequestAction | RequestFailAction | AuthenticateSuccessAction | LogoutSuccessAction
