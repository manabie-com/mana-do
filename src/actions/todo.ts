import { Todo } from "../models/todo";
import { ActionTypes } from "../constants";
import { Action } from "./commons";

export function setTodos(todos: Array<Todo>): Action<Array<Todo>> {
    return {
        type: ActionTypes.SET_TODO,
        payload: todos
    }
}


export function createTodo(newTodo: Todo): Action<Todo> {
    return {
        type: ActionTypes.CREATE_TODO,
        payload: newTodo
    }
}

export function updateTodoStatus(todoId: string, checked: boolean): Action<{ todoId: string, checked: boolean }> {
    return {
        type: ActionTypes.UPDATE_TODO_STATUS,
        payload: {
            todoId,
            checked
        }
    }
}

export function deleteTodo(todoId: string): Action<string> {
    return {
        type: ActionTypes.DELETE_TODO,
        payload: todoId
    }
}


export function deleteAllTodos(): Action<never> {
    return {
        type: ActionTypes.DELETE_ALL_TODOS,
    }
}

export function toggleAllTodos(checked: boolean): Action<boolean> {
    return {
        type: ActionTypes.TOGGLE_ALL_TODOS,
        payload: checked
    }
}