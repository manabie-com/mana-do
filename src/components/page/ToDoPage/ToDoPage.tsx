import React, { useEffect, useState } from 'react';

import {
  setTodos,
  toggleAllTodos,
  deleteAllTodos,
} from '../../../store/actions';
import Service from '../../../service';
import { TodoStatus } from '../../../models/todo';
import { isTodoCompleted } from '../../../utils';
import ToDoInput from '../../common/ToDoInput/ToDoInput.components';
import ToDoItem from '../../common/ToDoItem/ToDoItem.components';
import ActionButton from '../../common/ActionButton/ActionButton.components';
import ToDoTabs from '../../common/ToDoTabs/ToDoTabs.components';
import { useToDoPageContext } from '../../../context/ToDoPageProvider';

export type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = () => {
  const { todos, dispatch } = useToDoPageContext();
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  // create error message state for error handling
  const [error, setError] = React.useState<string>('');
  const [loading, setLoading] = React.useState<boolean>(false);
  useEffect(() => {
    setLoading(true);
    (async () => {
      try {
        const resp = await Service.getTodos();
        dispatch(setTodos(resp));
      } catch (error) {
        setError('Have an error when getTodos');
      }
    })();
    setLoading(false);
  }, [dispatch]);

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;
    // Optimis update
    Service.toggleAllTodos(checked);
    dispatch(toggleAllTodos(checked));
  };

  const onDeleteAllTodo = async () => {
    try {
      await Service.deleteAllTodos();
      dispatch(deleteAllTodos());
    } catch (error) {
      setError('Have error when delete all todos');
    }
  };

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  return (
    <div className='ToDo__container'>
      <div className='Todo__creation'>
        <ToDoInput />
      </div>
      {!loading ? (
        <>
          <div className='ToDo__list'>
            {!showTodos.length && <span>Nothing to do</span>}
            {showTodos.map((todo) => {
              const { id, content, created_date } = todo;
              return (
                <ToDoItem
                  key={id}
                  id={id}
                  checked={isTodoCompleted(todo)}
                  content={content}
                  createdDate={new Date(created_date)}
                />
              );
            })}
          </div>
          <div className='Todo__toolbar'>
            {todos.length > 0 ? (
              <input
                data-testid='toggle-all-todos'
                type='checkbox'
                checked={activeTodos === 0}
                onChange={onToggleAllTodo}
              />
            ) : (
              <div />
            )}
            <ToDoTabs setShowing={setShowing} activeTab={showing} />
            <ActionButton
              data-testid='clear-all-todos'
              className='Action__btn'
              onClick={onDeleteAllTodo}
            >
              Clear all todos
            </ActionButton>
          </div>
          {error && (
            <div>
              <p style={{ color: 'red' }}>{error}</p>
            </div>
          )}
        </>
      ) : (
        <span>Loading</span>
      )}
    </div>
  );
};

export default ToDoPage;
