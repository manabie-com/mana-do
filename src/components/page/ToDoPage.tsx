import React, { useEffect, useState } from 'react';

import { setTodos, toggleAllTodos, deleteAllTodos } from '../../store/actions';
import Service from '../../service';
import { TodoStatus } from '../../models/todo';
import { isTodoCompleted } from '../../utils';
import ToDoInput from '../common/ToDoInput/ToDoInput.components';
import ToDoItem from '../common/ToDoItem/ToDoItem.components';
import ActionButton from '../common/ActionButton/ActionButton.components';
import ToDoTabs from '../common/ToDoToolbar/ToDoTabs.components';
import { useToDoPageContext } from '../../context/ToDoPageProvider';

export type EnhanceTodoStatus = TodoStatus | 'ALL';

const ToDoPage = () => {
  const { todos, dispatch } = useToDoPageContext();
  const [showing, setShowing] = useState<EnhanceTodoStatus>('ALL');
  // create error message state for error handling
  const [error, setError] = React.useState<String>('');
  useEffect(() => {
    (async () => {
      try {
        const resp = await Service.getTodos();
        dispatch(setTodos(resp));
      } catch (error) {
        setError('Have an error when getTodos');
      }
    })();
  }, [dispatch]);

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { checked } = e.target;
    Service.toggleAllTodos(checked);
    dispatch(toggleAllTodos(checked));
  };

  const onDeleteAllTodo = async () => {
    await Service.deleteAllTodos();
    dispatch(deleteAllTodos());
  };

  const showTodos = todos.filter((todo) => {
    switch (showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const activeTodos = todos.reduce(function (accum, todo) {
    return isTodoCompleted(todo) ? accum : accum + 1;
  }, 0);

  return (
    <div className='ToDo__container'>
      <div className='Todo__creation'>
        <ToDoInput />
      </div>
      <div className='ToDo__list'>
        {showTodos.map((todo) => {
          const { id, content } = todo;
          return (
            <ToDoItem
              key={id}
              id={id}
              checked={isTodoCompleted(todo)}
              content={content}
            />
          );
        })}
      </div>
      <div className='Todo__toolbar'>
        {todos.length > 0 ? (
          <input
            data-testid='toggle-all-todos'
            type='checkbox'
            checked={activeTodos === 0}
            onChange={onToggleAllTodo}
          />
        ) : (
          <div />
        )}
        <ToDoTabs setShowing={setShowing} />
        <ActionButton
          data-testid='clear-all-todos'
          className='Action__btn'
          onClick={onDeleteAllTodo}
        >
          Clear all todos
        </ActionButton>
      </div>
      {error && (
        <div>
          <p style={{ color: 'red' }}>{error}</p>
        </div>
      )}
    </div>
  );
};

export default ToDoPage;
