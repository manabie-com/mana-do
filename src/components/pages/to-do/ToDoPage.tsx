import { useEffect, useReducer, useState } from "react";

import reducer, { initialState } from "../../../store/reducer";
import { setTodos } from "../../../store/actions";
import Service from "../../../service";
import { TodoStatus } from "../../../models/todo";
import ToolBar from "../../common/tool-bar/ToolBar";
import TaskCreateInput from "../../common/to-do-create-input/ToDoCreateInput";
import ToDoList from "../../common/to-do-list/ToDoList";

type EnhanceTodoStatus = TodoStatus | "ALL";

const ToDoPage = () => {
    const token = localStorage.getItem('token');
    if (!token) {
        window.location.href = 'login';
    }
    const [{ todos }, dispatch] = useReducer(reducer, initialState);
    const [showing, setShowing] = useState<EnhanceTodoStatus>("ALL");

    useEffect(() => {
        (async () => {
            const res = await Service.getTodos();
            dispatch(setTodos(res.data || []));
        })();
    }, []);

    const toolBarProps = {
        todos,
        dispatch,
        setShowing,
    };
    const taskCreateInputProps = {
        dispatch,
    };
    const toDoListProps = {
        todos,
        dispatch,
        showing,
    };
    return (
        <div className="to-do__container">
            <div className="to-do__creation">
                <TaskCreateInput {...taskCreateInputProps} />
            </div>
            <div className="to-do__toolbar">
                <ToolBar {...toolBarProps} />
            </div>
            <div className="to-do__list">
                <ToDoList {...toDoListProps} />
            </div>
        </div>
    );
};

export default ToDoPage;
