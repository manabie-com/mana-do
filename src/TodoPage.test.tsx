import * as React from 'react';
import renderer from 'react-test-renderer';
import {render, unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import {fireEvent} from '@testing-library/react';

import * as AppContext from './AppContext';
import * as ApiService from './service/api-frontend';
import * as actions from './store/actions';
import ToDoPage from "./ToDoPage";
import {TodoStatus} from "./models/todo";

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
}))
  .mock('./features/footer/Footer', () => 'Footer')
  .mock('./features/todos/TodoList', () => 'TodoList')
  .mock('./Layout', () => 'Layout')

let container: any = null;
const mockDispatch = jest.fn();

beforeEach(() => {
  const mockContextValue = {
    state: {
      todos: []
    },
    dispatch: mockDispatch
  }
  jest
    .spyOn(AppContext, 'useAppContext')
    // @ts-ignore
    .mockImplementation(() => mockContextValue);

  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('snapshot', () => {
  test('should return snapshot', () => {
    const tree = renderer.create(<ToDoPage/>).toJSON();
    expect(tree).toMatchSnapshot();
  })
})

describe('fetch todo list', () => {
  test('should fetch todo list success', async () => {
    const mockedTodos = [
      {
        id: 'hfuwyr',
        content: 'todo 1',
        user_id: 'firstUser',
        status: TodoStatus.Active,
        created_date: ''
      },
      {
        id: 'orislaj',
        content: 'todo 2',
        user_id: 'firstUser',
        status: TodoStatus.Active,
        created_date: ''
      },
    ]
    ApiService.default.getTodos = jest.fn().mockImplementation(() => Promise.resolve(mockedTodos));

    const setTodos = jest.spyOn(actions, 'setTodos');

    await act(async () => {
      render(<ToDoPage/>, container);
    });
    expect(setTodos).toHaveBeenCalledWith(mockedTodos);
  })

  test('should fetch todo list failed', async () => {
    const errorMessage = 'Fetch todos failed';
    ApiService.default.getTodos = jest.fn().mockImplementation(() => Promise.reject(errorMessage));

    await act(async () => {
      render(<ToDoPage/>, container);
    });
    expect(mockDispatch).toHaveBeenCalledWith({
      type: 'FETCH_TODO_LIST_FAILURE'
    });
  })
})

describe('create todo list', () => {
  const setTodosValue = jest.fn();
  const mockedTodos = [
    {
      id: 'hfuwyr',
      content: 'todo 1',
      user_id: 'firstUser',
      status: TodoStatus.Active,
      created_date: ''
    },
    {
      id: 'orislaj',
      content: 'todo 2',
      user_id: 'firstUser',
      status: TodoStatus.Active,
      created_date: ''
    },
  ]
  const useStateMock = () => [mockedTodos, setTodosValue];

  // @ts-ignore
  jest.spyOn(React, 'useState').mockImplementation(useStateMock);

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('should create todo list success', async () => {
    const mockTodoCreated = {
      id: 'hljfiu',
      content: 'todo 3',
      user_id: 'firstUser',
      status: TodoStatus.Active,
      created_date: ''
    }
    const expectedTodos = [{...mockTodoCreated}, ...mockedTodos];
    ApiService.default.createTodo = jest.fn().mockImplementation(() => Promise.resolve(mockTodoCreated));

    const createTodo = jest.spyOn(actions, 'createTodo');

    await act(async () => {
      render(<ToDoPage/>, container);
      const inputTask = container.querySelector('[class="Todo__input"]');
      fireEvent.keyDown(inputTask, {key: 'Enter', code: 'Enter'})
    });

    expect(createTodo).toHaveBeenCalledWith(mockTodoCreated);
    expect(setTodosValue).toHaveBeenCalledWith(expectedTodos);
  })

  test('should create todo list failed', async () => {
    const responseError = {
      response: {
        status: 401
      }
    }
    ApiService.default.createTodo = jest.fn().mockImplementation(() => Promise.reject(responseError));

    await act(async () => {
      render(<ToDoPage/>, container);
      const inputTask = container.querySelector('[class="Todo__input"]');
      fireEvent.keyDown(inputTask, {key: 'Enter', code: 'Enter'})
    });
  //  TODO: expect history push to sign in page
  })
})
