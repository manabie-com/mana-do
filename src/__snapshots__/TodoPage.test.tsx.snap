// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should return snapshot 1`] = `
<Layout>
  <section
    className="ToDo__container"
  >
    <h2>
      Todos
    </h2>
    <div
      className="ToDo__app"
    >
      <div
        className="Todo__creation"
      >
        <svg
          className="Todo__creation_icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
          />
        </svg>
        <input
          className="Todo__input"
          onKeyDown={[Function]}
          placeholder="Add a Task"
        />
      </div>
      <TodoList />
      <Footer />
    </div>
  </section>
</Layout>
`;
